Sources/Imperial/Services/Auth0/Auth0Router.swift

```diff
@@ -0,0 +1,90 @@
+import Vapor
+import Foundation
+
+public class Auth0Router: FederatedServiceRouter {
+    public let baseURL: String
+    public let tokens: FederatedServiceTokens
+    public let callbackCompletion: (Request, String)throws -> (Future<ResponseEncodable>)
+    public var scope: [String] = [ ]
+    public var requiredScopes = [ "openid" ]
+    public let callbackURL: String
+    public let accessTokenURL: String
+
+    private func providerUrl(path: String) -> String {
+        return self.baseURL.finished(with: "/") + path
+    }
+    
+    public required init(callback: String, completion: @escaping (Request, String)throws -> (Future<ResponseEncodable>)) throws {
+        let auth = try Auth0Auth()
+        self.tokens = auth
+        self.baseURL = "https://\(auth.domain)"
+        self.accessTokenURL = baseURL.finished(with: "/") + "oauth/token"
+        self.callbackURL = callback
+        self.callbackCompletion = completion
+    }
+    
+    public func authURL(_ request: Request) throws -> String {
+        let path="authorize"
+        let scopes = self.scope + self.requiredScopes
+        let scopeString = scopes.joined(separator: " ").addingPercentEncoding(withAllowedCharacters: .urlHostAllowed)!
```

Comment:
We shouldn't be force unwrapping here

Reply:
In this case I knew that requiredScopes always had at least 1 member, so it would be safe.  But I agree about avoiding force unwraps.  Fixed.